File: julia/test/test_roundtrip.jl
#!/usr/bin/env julia

# Test MPC designation with bidirectional timing and round-trip verification

push!(LOAD_PATH, joinpath(@__DIR__, "..", "src"))

using MPCDesignation

struct TestCase
    unpacked::String
    packed::String
end

function main()
    if length(ARGS) < 1
        println(stderr, "Usage: julia test_roundtrip.jl <prov_unpack_to_pack.csv>")
        exit(1)
    end

    csv_file = ARGS[1]
    test_cases = TestCase[]

    # Load test cases
    open(csv_file) do f
        is_first_line = true
        for line in eachline(f)
            isempty(line) && continue

            if is_first_line
                is_first_line = false
                if startswith(line, "unpacked") || occursin("designation", line)
                    continue
                end
            end

            parts = split(line, ',', limit=2)
            length(parts) != 2 && continue

            push!(test_cases, TestCase(parts[1], parts[2]))
        end
    end

    println("Loaded $(length(test_cases)) test cases")
    println()

    pack_passed = 0
    pack_failed = 0
    unpack_passed = 0
    unpack_failed = 0
    rt_unpacked_passed = 0
    rt_unpacked_failed = 0
    rt_packed_passed = 0
    rt_packed_failed = 0

    # Phase 1: Pack (unpacked -> packed)
    println("=== Phase 1: Pack (unpacked -> packed) ===")
    start_time = time()
    for tc in test_cases
        try
            result = pack(tc.unpacked)
            if result == tc.packed
                pack_passed += 1
            else
                pack_failed += 1
            end
        catch
            pack_failed += 1
        end
    end
    elapsed = (time() - start_time) * 1000
    rate = length(test_cases) / (elapsed / 1000)
    println("Passed: $pack_passed")
    println("Failed: $pack_failed")
    println("Time:   $(round(Int, elapsed))ms ($(round(rate, digits=1)) entries/sec)")
    println()

    # Phase 2: Unpack (packed -> unpacked)
    println("=== Phase 2: Unpack (packed -> unpacked) ===")
    start_time = time()
    for tc in test_cases
        try
            result = unpack(tc.packed)
            if result == tc.unpacked
                unpack_passed += 1
            else
                unpack_failed += 1
            end
        catch
            unpack_failed += 1
        end
    end
    elapsed = (time() - start_time) * 1000
    rate = length(test_cases) / (elapsed / 1000)
    println("Passed: $unpack_passed")
    println("Failed: $unpack_failed")
    println("Time:   $(round(Int, elapsed))ms ($(round(rate, digits=1)) entries/sec)")
    println()

    # Phase 3: Unpacked round-trip: unpack(pack(x)) = x
    println("=== Phase 3: Unpacked round-trip: unpack(pack(x)) = x ===")
    start_time = time()
    for tc in test_cases
        try
            packed = pack(tc.unpacked)
            back = unpack(packed)
            if back == tc.unpacked
                rt_unpacked_passed += 1
            else
                rt_unpacked_failed += 1
            end
        catch
            rt_unpacked_failed += 1
        end
    end
    elapsed = (time() - start_time) * 1000
    rate = length(test_cases) / (elapsed / 1000)
    println("Passed: $rt_unpacked_passed")
    println("Failed: $rt_unpacked_failed")
    println("Time:   $(round(Int, elapsed))ms ($(round(rate, digits=1)) entries/sec)")
    println()

    # Phase 4: Packed round-trip: pack(unpack(y)) = y
    println("=== Phase 4: Packed round-trip: pack(unpack(y)) = y ===")
    start_time = time()
    for tc in test_cases
        try
            unpacked_result = unpack(tc.packed)
            back = pack(unpacked_result)
            if back == tc.packed
                rt_packed_passed += 1
            else
                rt_packed_failed += 1
            end
        catch
            rt_packed_failed += 1
        end
    end
    elapsed = (time() - start_time) * 1000
    rate = length(test_cases) / (elapsed / 1000)
    println("Passed: $rt_packed_passed")
    println("Failed: $rt_packed_failed")
    println("Time:   $(round(Int, elapsed))ms ($(round(rate, digits=1)) entries/sec)")
    println()

    # Summary
    println("=== Summary ===")
    println("Pack:       $(pack_failed == 0 ? "PASS" : "FAIL ($pack_failed)")")
    println("Unpack:     $(unpack_failed == 0 ? "PASS" : "FAIL ($unpack_failed)")")
    println("Unpacked RT: $(rt_unpacked_failed == 0 ? "PASS" : "FAIL ($rt_unpacked_failed)")")
    println("Packed RT:   $(rt_packed_failed == 0 ? "PASS" : "FAIL ($rt_packed_failed)")")

    if pack_failed > 0 || rt_packed_failed > 0
        exit(1)
    end
end

main()

File: haskell/test/TestRoundtrip.hs
{-|
Module      : TestRoundtrip
Description : Test MPC designation with bidirectional timing and round-trip verification
-}

module Main where

import System.Environment (getArgs)
import System.Exit (exitFailure, exitSuccess)
import System.IO (hPutStrLn, stderr)
import Data.Time.Clock (getCurrentTime, diffUTCTime)
import Control.DeepSeq (deepseq)
import MPCDesignation

main :: IO ()
main = do
    args <- getArgs
    case args of
        [csvFile] -> runTests csvFile
        _ -> do
            hPutStrLn stderr "Usage: test_roundtrip <csv_file>"
            exitFailure

data TestCase = TestCase
    { tcUnpacked :: String
    , tcPacked :: String
    } deriving (Show)

runTests :: FilePath -> IO ()
runTests csvFile = do
    contents <- readFile csvFile
    let allLines = lines contents
        dataLines = drop 1 allLines  -- Skip header
        testCases = map parseLine dataLines

    let total = length testCases
    putStrLn $ "Loaded " ++ show total ++ " test cases"
    putStrLn ""

    -- Phase 1: Pack (unpacked -> packed)
    putStrLn "=== Phase 1: Pack (unpacked -> packed) ==="
    startTime <- getCurrentTime

    let packResults = map testPack testCases
        (packPassed, packFailed) = countResults packResults

    packPassed `deepseq` packFailed `deepseq` return ()

    endTime <- getCurrentTime
    let elapsed1 = realToFrac (diffUTCTime endTime startTime) * 1000 :: Double
        rate1 = fromIntegral total / (elapsed1 / 1000)

    putStrLn $ "Passed: " ++ show packPassed
    putStrLn $ "Failed: " ++ show packFailed
    putStrLn $ "Time:   " ++ show (round elapsed1 :: Int) ++ "ms (" ++ formatRate rate1 ++ " entries/sec)"
    putStrLn ""

    -- Phase 2: Unpack (packed -> unpacked)
    putStrLn "=== Phase 2: Unpack (packed -> unpacked) ==="
    startTime2 <- getCurrentTime

    let unpackResults = map testUnpack testCases
        (unpackPassed, unpackFailed) = countResults unpackResults

    unpackPassed `deepseq` unpackFailed `deepseq` return ()

    endTime2 <- getCurrentTime
    let elapsed2 = realToFrac (diffUTCTime endTime2 startTime2) * 1000 :: Double
        rate2 = fromIntegral total / (elapsed2 / 1000)

    putStrLn $ "Passed: " ++ show unpackPassed
    putStrLn $ "Failed: " ++ show unpackFailed
    putStrLn $ "Time:   " ++ show (round elapsed2 :: Int) ++ "ms (" ++ formatRate rate2 ++ " entries/sec)"
    putStrLn ""

    -- Phase 3: Unpacked round-trip: unpack(pack(x)) = x
    putStrLn "=== Phase 3: Unpacked round-trip: unpack(pack(x)) = x ==="
    startTime3 <- getCurrentTime

    let rtUnpackedResults = map testUnpackedRoundtrip testCases
        (rtUnpackedPassed, rtUnpackedFailed) = countResults rtUnpackedResults

    rtUnpackedPassed `deepseq` rtUnpackedFailed `deepseq` return ()

    endTime3 <- getCurrentTime
    let elapsed3 = realToFrac (diffUTCTime endTime3 startTime3) * 1000 :: Double
        rate3 = fromIntegral total / (elapsed3 / 1000)

    putStrLn $ "Passed: " ++ show rtUnpackedPassed
    putStrLn $ "Failed: " ++ show rtUnpackedFailed
    putStrLn $ "Time:   " ++ show (round elapsed3 :: Int) ++ "ms (" ++ formatRate rate3 ++ " entries/sec)"
    putStrLn ""

    -- Phase 4: Packed round-trip: pack(unpack(y)) = y
    putStrLn "=== Phase 4: Packed round-trip: pack(unpack(y)) = y ==="
    startTime4 <- getCurrentTime

    let rtPackedResults = map testPackedRoundtrip testCases
        (rtPackedPassed, rtPackedFailed) = countResults rtPackedResults

    rtPackedPassed `deepseq` rtPackedFailed `deepseq` return ()

    endTime4 <- getCurrentTime
    let elapsed4 = realToFrac (diffUTCTime endTime4 startTime4) * 1000 :: Double
        rate4 = fromIntegral total / (elapsed4 / 1000)

    putStrLn $ "Passed: " ++ show rtPackedPassed
    putStrLn $ "Failed: " ++ show rtPackedFailed
    putStrLn $ "Time:   " ++ show (round elapsed4 :: Int) ++ "ms (" ++ formatRate rate4 ++ " entries/sec)"
    putStrLn ""

    -- Summary
    putStrLn "=== Summary ==="
    putStrLn $ "Pack:       " ++ (if packFailed == 0 then "PASS" else "FAIL (" ++ show packFailed ++ ")")
    putStrLn $ "Unpack:     " ++ (if unpackFailed == 0 then "PASS" else "FAIL (" ++ show unpackFailed ++ ")")
    putStrLn $ "Unpacked RT: " ++ (if rtUnpackedFailed == 0 then "PASS" else "FAIL (" ++ show rtUnpackedFailed ++ ")")
    putStrLn $ "Packed RT:   " ++ (if rtPackedFailed == 0 then "PASS" else "FAIL (" ++ show rtPackedFailed ++ ")")

    if packFailed > 0 || rtPackedFailed > 0
        then exitFailure
        else exitSuccess

parseLine :: String -> TestCase
parseLine line =
    let (unpacked, rest) = break (== ',') line
        packed = drop 1 rest
    in TestCase unpacked packed

testPack :: TestCase -> Bool
testPack tc =
    case pack (tcUnpacked tc) of
        Right result -> result == tcPacked tc
        Left _ -> False

testUnpack :: TestCase -> Bool
testUnpack tc =
    case unpack (tcPacked tc) of
        Right result -> result == tcUnpacked tc
        Left _ -> False

testUnpackedRoundtrip :: TestCase -> Bool
testUnpackedRoundtrip tc =
    case pack (tcUnpacked tc) of
        Right packed ->
            case unpack packed of
                Right back -> back == tcUnpacked tc
                Left _ -> False
        Left _ -> False

testPackedRoundtrip :: TestCase -> Bool
testPackedRoundtrip tc =
    case unpack (tcPacked tc) of
        Right unpacked ->
            case pack unpacked of
                Right back -> back == tcPacked tc
                Left _ -> False
        Left _ -> False

countResults :: [Bool] -> (Int, Int)
countResults = foldr count (0, 0)
  where
    count True (p, f) = (p + 1, f)
    count False (p, f) = (p, f + 1)

formatRate :: Double -> String
formatRate r = show (round r :: Int) ++ ".0"
